Overload Library::def rather than templating it

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #151682
* #151697
* #151630
* #151629
* #151628
* #151627
* __->__ #151626

It ends up being templated over a bunch of reference-to-array-of-characters types with different lengths, such as `char const (&) [88]`, which is an annoyance when profiling and possibly a source of code bloat.

Differential Revision: [D73129450](https://our.internmc.facebook.com/intern/diff/D73129450/)

cc @jbschlosser

swolchok committed:
Overload Library::def rather than templating it

It ends up being templated over a bunch of reference-to-array-of-characters types with different lengths, such as `char const (&) [88]`, which is an annoyance when profiling and possibly a source of code bloat.

Differential Revision: [D73129450](https://our.internmc.facebook.com/intern/diff/D73129450/)

[ghstack-poisoned]

albanD reviewed (commented):
I would have two questions:
- Any impact on import time? Since we have so many of these defs
- There area a lot of calls to this function from out of core, is this change BC-breaking for our C++ API?

malfet commented on torch/library.h:
```diff
@@ -605,12 +605,18 @@ class TORCH_API Library final {
-  template <typename Schema>
   Library& def(
-      Schema&& raw_schema,
+      c10::FunctionSchema s,
```
Why not pass it by reference if we expect it to be moved anyway?

swolchok replied:
pass-by-value this is a succinct way to implement "move if we can, copy if we have to". however, I checked the C++ core guidelines, and it appears they just [recommend](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f15-prefer-simple-and-conventional-ways-of-passing-information) the more efficient approach of having two overloads, one for const ref and one for rvalue ref, so I'll do that.

swolchok replied:
turns out this wouldn't have built in the first place with `const c10::FunctionSchema&`, so we can just use rvalue ref directly.

malfet reviewed (approved):
Seems fine, but let's slap a BC-breaking label, as it could potentially result in some failures.
And if those changes are about perf, would be nice to include some before/after stats on either binary size or import time

swolchok commented:
Re: bc breaking, can you elaborate on the argument types that you think worked before this PR and won't work after? I think it should be fine and if you can point to a type I missed, we can just cover it

swolchok commented:
> if those changes are about perf, would be nice to include some before/after stats on either binary size or import time

the motivation here was convenience while profiling; it's annoying to see stack traces fan out into a bunch of different instantiations of def. That said, I had expected that def should've been getting boiled away by inlining, but I checked assembly for RegisterSchema.cpp in OSS PyTorch build (just `python setup.py develop`, which seems to end up using `-O2` despite a mention of `-O3`) on my Mac, and apparently instead `_def` was getting inlined into `def`. The size improvement was roughly 85500 bytes in just RegisterSchema.cpp.o.

swolchok committed:
Update on "Overload Library::def rather than templating it"


It ends up being templated over a bunch of reference-to-array-of-characters types with different lengths, such as `char const (&) [88]`, which is an annoyance when profiling and possibly a source of code bloat.

Differential Revision: [D73129450](https://our.internmc.facebook.com/intern/diff/D73129450/)

cc jbschlosser

[ghstack-poisoned]

albanD commented:
> Re: bc breaking, can you elaborate on the argument types that you think worked before this PR and won't work after? I think it should be fine and if you can point to a type I missed, we can just cover it.

I don't have a concrete example off the top of my head. But the standard approach for BC-breaking change we have is that the PR author has to prove that everything that was possible before is still possible, not for the reviewer to come up with example where it breaks.
This is because we don't control call sites, and quite a lot of them are not even public code. So we have to consider the worst case scenario all the time to avoid breaking users.

If you're saying only these two types worked before and they still work after this change, that sounds good to me!
But then we don't need the bc-breaking label?
